fix: streaming finish_reason returns tool_calls for o1-pro and Cohere models

colesmcintosh opened the pull request:
## Title

Fix streaming finish_reason to return "tool_calls" instead of "stop" for o1-pro and Cohere models when using function tools

## Relevant issues

Fixes #12862

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type

🐛 Bug Fix

## Changes

When using streaming with function tools, the final chunk was returning `"finish_reason": "stop"` instead of `"finish_reason": "tool_calls"` for o1-pro and Cohere models. This broke agentic workflows that rely on detecting tool call completions.

### What was changed:
1. Modified `streaming_handler.py` to detect when tool_calls or function_call are present in streaming deltas
2. Set `self.tool_call = True` flag when tool calls are detected in OpenAI-style streaming responses
3. Override finish_reason from "stop" to "tool_calls" when the tool_call flag is set

### Test coverage:
Added comprehensive unit tests in `test_streaming_finish_reason_tool_calls.py` that verify:
- OpenAI streaming with tool_calls gets correct finish_reason
- Generic streaming (Cohere/Anthropic style) with tool_use gets correct finish_reason  
- Legacy function_call format also gets correct finish_reason
- Regular messages without tool calls keep "stop" finish_reason

### Test results:
All tests pass successfully:
```
pytest tests/test_litellm/test_streaming_finish_reason_tool_calls.py -v
```

colesmcintosh committed:
fix: streaming finish_reason returns tool_calls for o1-pro and Cohere models

When using streaming with function tools, the final chunk now correctly returns
finish_reason="tool_calls" instead of "stop" for o1-pro and Cohere models.

This fix ensures consistent behavior across all providers by:
- Setting tool_call flag when tool_calls or function_call are detected in delta
- Overriding finish_reason from "stop" to "tool_calls" when tool calls are present

Fixes #12862

colesmcintosh committed:
fix: Cohere streaming responses now correctly return finish_reason tool_calls

- Updated chunk_parser in common_utils.py to handle tool_call_delta events
- Fixed SSE parsing to properly strip "data:" prefix
- Set finish_reason to "tool_calls" when tool calls are present in final chunk
- Added proper handling of streaming tool call chunks with ID and arguments

colesmcintosh committed:
test: Add tests for Cohere tool_calls finish_reason

- Tests verify that finish_reason is set to "tool_calls" when tool calls are present
- Tests both v1 and v2 Cohere transformation classes
- Ensures finish_reason remains "stop" when no tool calls are present
- Validates the fix works correctly with streaming_handler.py override

colesmcintosh committed:
fix: Cohere streaming now tracks tool calls state for correct finish_reason

- Added state tracking to ModelResponseIterator to remember if tool calls were seen
- Sets finish_reason to "tool_calls" in final chunk if any tool calls were present
- Fixes issue where Cohere sends tool calls in intermediate chunks but not final chunk
- Avoids need for global streaming_handler override

colesmcintosh committed:
test: Remove streaming_handler tests as fix is now Cohere-specific

Since we're fixing the issue in Cohere's transformation code rather than
globally in streaming_handler, we don't need the generic streaming tests.
The Cohere-specific tests remain to verify the fix works correctly.

krrishdholakia commented on litellm/litellm_core_utils/streaming_handler.py:
please add a unit test for the streaming_handler

colesmcintosh replied:
I've completed manual QA. I tested streaming responses using both OpenAI (gpt-4o-mini) and Cohere (command-r-plus) models across two scenarios:

* ✅ With tool calls: Both models now correctly return `finish_reason: "tool_calls"`
* ✅ Without tool calls: Both models correctly return `finish_reason: "stop"`

This confirms the fix resolves the issue where streaming responses with tool calls were incorrectly returning `"stop"` instead of `"tool_calls"`, while preserving expected behavior for normal responses.

I've also added unit tests for the `streaming_handler`

colesmcintosh committed:
Merge branch 'BerriAI:main' into fix/streaming-finish-reason-tool-calls

colesmcintosh committed:
test: add unit tests for streaming_handler tool call finish_reason

Added comprehensive unit tests to verify that streaming_handler correctly
sets finish_reason to "tool_calls" when tool calls are present in the
streaming response. Tests cover:

- OpenAI-style tool_calls format
- Legacy function_call format
- Regular messages without tool calls (finish_reason remains "stop")
- Other finish reasons like "length" are not modified

Manual QA confirmed the fix works correctly with:
- OpenAI GPT-4o-mini: ✅ Returns "tool_calls" with tools, "stop" without
- Cohere Command R+: ✅ Returns "tool_calls" with tools, "stop" without

colesmcintosh committed:
Merge branch 'main' into fix/streaming-finish-reason-tool-calls

Resolved conflict in litellm/litellm_core_utils/streaming_handler.py by combining both changes:
- Kept the tool_call flag logic from the PR branch
- Used the new strip_role_from_delta method from main branch
